developer-guide.md

```diff
@@ -0,0 +1,184 @@
+  grep -E " Running |\[ERROR\]|Failures|SUCCESS|Skipp|Total time|Finished"
+
+[INFO] Running com.ibm.watson.modelmesh.ModelMeshFailureExpiryTest
+[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 10.257 s - in com.ibm.watson.modelmesh.ModelMeshFailureExpiryTest
+[INFO] Running com.ibm.watson.modelmesh.SidecarModelMeshTest
+[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 17.302 s - in com.ibm.watson.modelmesh.SidecarModelMeshTest
+[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
+[INFO] BUILD SUCCESS
+[INFO] Total time:  39.916 s
+[INFO] Finished at: 2022-11-01T14:33:33-07:00
+```
+
+## Building the container image
+
+Sample build:
+
+```bash
+export DOCKER_USER=$(
+  docker-credential-$(jq -r .credsStore ~/.docker/config.json) list | \
+  jq -r '. | to_entries[] | select(.key | contains("docker.io")) | last(.value)' | head -1
+)
+export IMAGE_TAG="dev" # "v0.10.0", "latest"
+export IMAGE_NAME="${DOCKER_USER}/modelmesh"
+export GIT_COMMIT=$(git rev-parse HEAD)
+export BUILD_ID=$(date '+%Y%m%d')-$(git rev-parse HEAD | cut -c -5)
+
+docker build -t ${IMAGE_NAME}:${IMAGE_TAG} \
+    --build-arg imageVersion=${IMAGE_TAG} \
+    --build-arg buildId=${BUILD_ID} \
+    --build-arg commitSha=${GIT_COMMIT} .
+
+docker push ${IMAGE_NAME}:${IMAGE_TAG}
+```
+
+## Updating the ModelMesh deployment
+
+In order to test your code changes in an existing ModelMesh deployment, you need to add the
+container image you built earlier to your `model-serving-config` configmap.
+
+If your ModelMesh deployment already has an existing `model-serving-config` ConfigMap,
+save the contents in a local temp file:
+
+```Bash
+mkdir -p temp
+kubectl get configmap model-serving-config -o yaml > temp/model-serving-config.yaml
+```
+
+And add the `modelMeshImage` property to the `config.yaml` string property in the temp file:
+```YAML
+      modelMeshImage:
+        name: yourdockeruser/modelmesh
+        tag: dev
+```
+
+Replace the `yourdockeruser` placeholder with your Docker username/login.
+
+The complete ConfigMap YAML file might look like this:
+```YAML
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: model-serving-config
+  namespace: modelmesh-serving
+data:
+  config.yaml: |
+    podsPerRuntime: 1
+    restProxy:
+      enabled: true
+    scaleToZero:
+      enabled: false
+      gracePeriodSeconds: 5
+    modelMeshImage:
+      name: yourdockeruser/modelmesh
+      tag: dev
+```
+
+Apply the ConfigMap to your cluster:
+
+```Bash
+kubectl apply -f temp/model-serving-config.yaml
+```
+
+If you did not already have a `model-serving-config` ConfigMap on your cluster, you can create one like this:
+
+```shell
+kubectl apply -f - <<EOF
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: model-serving-config
+data:
+  config.yaml: |
+    modelMeshImage:
+      name: yourdockeruser/modelmesh
+      tag: dev
+EOF
+```
+
+For the changes to take effect, you might have to delete existing `modelmesh-controller` pod.
```

Comment:
Even after deleting the pod, restarted pod did not pick up the new image

Reply:
In theory even a restart of the controller pod shouldn't be required since it watches the configmap and responds to updates. I guess it would be a good exercise to investigate/debug this if it's not happening.

Reply:
In my last few attempts, the runtime pods were restarted automatically after updating the `model-serving-config` ConfigMap.

I think the problems come with editing the `model-serving-config-defaults` instead of creating (or modifying) the `model-serving-config` config map.